src/scf/newton.jl

```diff
@@ -0,0 +1,489 @@
+                        # important to specify custom orthogonaliser to keep
+                        # iterations on the tangent space
+                        orth=OrthogonalizeAndProject(packed_proj, pack(φproj)))
+    δφ = unpack(δφ)
+    # ensure proper projection onto the tangent space
+    δφ = proj_tangent(δφ, φ)
+    φ_newton = similar(φ)
+
+    # perform newton_step
+    for ik = 1:Nk
+        φ_newton[ik] = ortho_qr(φ[ik] - δφ[ik])
+    end
+    (φ_newton, δφ)
+end
+
+"""
+    newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+                    tol=1e-6, maxiter=20, φproj=nothing, verbosity=0,
+                    callback=NewtonDefaultCallback(),
+                    is_converged=NewtonConvergenceDensity(tol))
+
+Newton algorithm. Be careful that the starting needs to be not too far from the solution.
+Still contains debugging features.
+If φproj is nothing, we use the successive φ's to define the tangent spaces on which we solve
+Newton's equation. φproj can be useful when we want to fix the tangent space for testing.
+"""
+function newton(basis::PlaneWaveBasis{T}; ψ0=nothing,
+                tol=1e-6, maxiter=20, φproj=nothing, verbosity=0,
+                callback=NewtonDefaultCallback(),
+                is_converged=NewtonConvergenceDensity(tol)) where T
+
+    ## setting parameters
+    model = basis.model
+    @assert model.spin_polarization in (:none, :spinless, :collinear)
+    @assert model.temperature == 0 # temperature is not yet supported
+    filled_occ = filled_occupation(model)
+    n_spin = basis.model.n_spin_components
+    N = div(div(model.n_electrons, filled_occ), n_spin)
+
+    ## number of kpoints and occupation
+    Nk = length(basis.kpoints)
+    occupation = [filled_occ * ones(T, N) for ik = 1:Nk]
+
+    ## starting point and keep only occupied orbitals
+    if ψ0 === nothing
+        ψ0 = [ortho_qr(randn(Complex{T}, length(G_vectors(kpt)), N))
+              for kpt in basis.kpoints]
+    else
+        for ik in 1:Nk
+            ψ0[ik] = ψ0[ik][:, 1:N]
+        end
+    end
+    update_φproj = true
+    if φproj != nothing
+        update_φproj = false
+        for ik in 1:Nk
+            φproj[ik] = φproj[ik][:, 1:N]
+        end
+    end
+
+    ## iterators
+    err = 1
+    n_iter = 0
+
+    ## orbitals, densities and energies to be updated along the iterations
+    φ = deepcopy(ψ0)
+    ρ = compute_density(basis, φ, occupation)
+    prev_energies = nothing
+    converged = false
+
+    ## perform iterations
+    while !converged && n_iter < maxiter
+        n_iter += 1
+
+        # update φproj which defines the tangent space if necessary
+        if update_φproj
+            φproj = φ
+        end
+
+        # compute next step
+        res = compute_scf_residual(basis, φ, occupation)
+        φ, δφ = newton_step(basis, φ, φproj, res, occupation;
+                            krylov_verbosity=verbosity)
+
+        # callback
+        ρ_next = compute_density(basis, φ, occupation)
```

Comment:
That's a bit wasteful (rho and H will need to be computed twice per step), but we don't really care that much

Reply:
I can add rho and H as argument of newton_step to avoid that if you want.